python/semantic_kernel/agents/orchestration/magentic.py

```diff
@@ -0,0 +1,854 @@
+
+    max_stall_count: Annotated[int, Field(description="The maximum number of stalls allowed before a reset.", ge=0)] = 3
+    max_reset_count: Annotated[int | None, Field(description="The maximum number of resets allowed.", ge=0)] = None
+    max_round_count: Annotated[
+        int | None, Field(description="The maximum number of rounds (agent responses) allowed.", gt=0)
+    ] = None
+
+    @abstractmethod
+    async def plan(self, magentic_context: MagenticContext) -> ChatMessageContent:
+        """Create a plan for the task.
+
+        This is called when the task is first started.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ChatMessageContent: The task ledger.
+        """
+        ...
+
+    @abstractmethod
+    async def replan(self, magentic_context: MagenticContext) -> ChatMessageContent:
+        """Replan for the task.
+
+        This is called when the task is stalled or looping.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ChatMessageContent: The updated task ledger.
+        """
+        ...
+
+    @abstractmethod
+    async def create_progress_ledger(self, magentic_context: MagenticContext) -> ProgressLedger:
+        """Create a progress ledger.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ProgressLedger: The progress ledger.
+        """
+        ...
+
+    @abstractmethod
+    async def prepare_final_answer(self, magentic_context: MagenticContext) -> ChatMessageContent:
+        """Prepare the final answer.
+
+        Args:
+            magentic_context (MagenticContext): The context for the Magentic manager.
+
+        Returns:
+            ChatMessageContent: The final answer.
+        """
+        ...
+
+
+@experimental
+class StandardMagenticManager(MagenticManager):
+    """Standard Magentic manager implementation.
+
+    This is the default implementation of the Magentic manager.
+    It uses the task ledger to keep track of the facts and plan for the task.
+
+    This implementation is requires structured outputs.
+    """
+
+    chat_completion_service: ChatCompletionClientBase
+    prompt_execution_settings: PromptExecutionSettings
+
+    task_ledger_facts_prompt: str = ORCHESTRATOR_TASK_LEDGER_FACTS_PROMPT
+    task_ledger_plan_prompt: str = ORCHESTRATOR_TASK_LEDGER_PLAN_PROMPT
+    task_ledger_full_prompt: str = ORCHESTRATOR_TASK_LEDGER_FULL_PROMPT
+    task_ledger_facts_update_prompt: str = ORCHESTRATOR_TASK_LEDGER_FACTS_UPDATE_PROMPT
+    task_ledger_plan_update_prompt: str = ORCHESTRATOR_TASK_LEDGER_PLAN_UPDATE_PROMPT
+    progress_ledger_prompt: str = ORCHESTRATOR_PROGRESS_LEDGER_PROMPT
+    final_answer_prompt: str = ORCHESTRATOR_FINAL_ANSWER_PROMPT
+
+    class TaskLedger(KernelBaseModel):
```

Comment:
I'm not a fan of classes within classes, so can't this just be defined above, next to the ProcessLedger?